### PR TITLE
[COH-18373] Enhance/trace levels and context

### PR DESCRIPTION
**Pull request by palikar**

### CODE REVIEW

**Review by alien1976 — APPROVED** [review]
Works perfectly!

Suggestions for feature improvements:
1. Multiselect as we discussed in the thread -> https://coherentlabs.slack.com/archives/C7CEWG1CL/p1704357289675799
2. Can we hide the options when `Trace Filtering Enabled` is disabled.
    2.1. Also can we rename the option to `Trace Filtering` if we hide the options.
3. `Trace level` options are not descriptive enough for me but could be for the others. However maybe they will be more descriptive if are named `Level 1`, `Level 2`, `Level 3`
4. Fix the names in the `Trace System` to be `Resource loading` instead of `ResourceLoading`
5. Add `:` to the options so they are consistent with the rest Inspector options -> for example `Trace level:` or `Trace system:`

image

**Comment by palikar** [comment]
1. I've created a JIRA [task](https://coherent-labs.atlassian.net/browse/COH-18770) for that.
2. Renamed the field to "Trace Filtering"
3. The levels are now "Level 1/2/3"
4. Fixed the names (this was done on the backend)
5. Added ":"

File: front_end/generated/InspectorBackendCommands.js [hunk 1]
```diff
@@ -752,6 +752,7 @@ inspectorBackend.registerEnum("Tracing.TracingBackend", {Auto: "auto", Chrome: "
 inspectorBackend.registerEvent("Tracing.bufferUsage", ["percentFull", "eventCount", "value"]);
 inspectorBackend.registerEvent("Tracing.dataCollected", ["value"]);
 inspectorBackend.registerEvent("Tracing.tracingComplete", ["dataLossOccurred", "stream", "traceFormat", "streamCompression"]);
+inspectorBackend.registerCommand("Tracing.getTraceSystemsAndLevels", [], ["systems", "levels"]);
```
Comment (omalinov) [inline front_end/generated/InspectorBackendCommands.js:755]:
```suggestion
// COHERENT BEGIN
inspectorBackend.registerCommand("Tracing.getTraceSystemsAndLevels", [], ["systems", "levels"]);
// COHERENT END
```

File: front_end/generated/protocol-mapping.d.ts [hunk 2]
```diff
@@ -3546,6 +3546,10 @@ export namespace ProtocolMapping {
       paramsType: [Protocol.Tethering.UnbindRequest];
       returnType: void;
     };
+    'Tracing.getTraceSystemsAndLevels': {
+      paramsType: [];
+      returnType: Protocol.Tracing.GetTraceSystemsAndLevelsResponse;
+    };
```
Comment (omalinov) [inline front_end/generated/protocol-mapping.d.ts:3552]:
```suggestion
    // COHERENT BEGIN
    'Tracing.getTraceSystemsAndLevels': {
      paramsType: [];
      returnType: Protocol.Tracing.GetTraceSystemsAndLevelsResponse;
    };
    // COHERENT END
```

File: front_end/generated/protocol-proxy-api.d.ts [hunk 3]
```diff
@@ -3053,6 +3053,8 @@ declare namespace ProtocolProxyApi {
   }
 
   export interface TracingApi {
+    invoke_getTraceSystemsAndLevels(): Promise<Protocol.Tracing.GetTraceSystemsAndLevelsResponse>;
```
Comment (omalinov) [inline front_end/generated/protocol-proxy-api.d.ts:3056]:
```suggestion
    // COHERENT BEGIN
    invoke_getTraceSystemsAndLevels(): Promise<Protocol.Tracing.GetTraceSystemsAndLevelsResponse>;
    // COHERENT END
```

File: front_end/generated/protocol.d.ts [hunk 4]
```diff
@@ -12993,6 +12993,19 @@ declare namespace Protocol {
...
+     */
+    export interface TraceCategoryDefinition {
+      name: string;
+      categoryString: string;
+    }
+
+    export interface GetTraceSystemsAndLevelsResponse extends ProtocolResponseWithError {
+      systems: TraceCategoryDefinition[];
+      levels: TraceCategoryDefinition[];
+    }
```
Comment (omalinov) [inline front_end/generated/protocol.d.ts:13007]:
```suggestion
    // COHERENT BEGIN
    export interface TraceCategoryDefinition {
      name: string;
      categoryString: string;
    }

    export interface GetTraceSystemsAndLevelsResponse extends ProtocolResponseWithError {
      systems: TraceCategoryDefinition[];
      levels: TraceCategoryDefinition[];
    }
    // COHERENT END
```

File: front_end/panels/timeline/TimelinePanel.ts [hunk 5]
```diff
@@ -420,6 +460,55 @@ export class TimelinePanel extends UI.Panel.Panel implements Client, TimelineMod
...
+    } else {
+      this._controller = new TimelineController(mainTarget, this);
+    }
+
+    this._controller?._tracingManager?.fetchTraceSystemsAndLevels().then((res) => {
+      this._traceSystemsDefinitions = res.systems;
+      this._traceLevelsDefinitions = res.levels;
+
+      this._controller?.dispose();
+      this._controller = null;
```
Comment (omalinov) [inline front_end/panels/timeline/TimelinePanel.ts]: This should break the world when you receive the response while tracing - e.g. when you refresh and trace a page (not sure if it occurs, but seems possible.

What if instead we get the tracing agent directly, as we seem to construct the controller for it.
Maybe through those calls https://github.com/CoherentLabs/devtools-frontend-fork/blob/13706efede840af5ccdc8ed8d62b38d4d310ab63/front_end/core/protocol_client/InspectorBackend.ts#L733

The following seems okay at a first glance, do you find issues with it?
```JS
fetchTraceSystemsAndLevels(agent) {
    const response = await agent.invoke_getTraceSystemsAndLevels();
    return response;
}

fetchTraceSystemsAndLevels(tracingAgent()).then(res => {
  // setup stuff
});
```

File: front_end/generated/InspectorBackendCommands.js (hunk 1 above)
Comment (palikar) [reply front_end/generated/InspectorBackendCommands.js:755]: this is a generated file. All `COHERENT BEGIN` / `COHERENT END` comments for changes responsible for this are in `browser_protocol.pdl`

File: front_end/generated/protocol-mapping.d.ts (hunk 2 above)
Comment (palikar) [reply front_end/generated/protocol-mapping.d.ts:3552]: this is a generated file. All `COHERENT BEGIN` / `COHERENT END` comments for changes responsible for this are in `browser_protocol.pdl`

File: front_end/generated/protocol-proxy-api.d.ts (hunk 3 above)
Comment (palikar) [reply front_end/generated/protocol-proxy-api.d.ts:3056]: this is a generated file. All `COHERENT BEGIN` / `COHERENT END` comments for changes responsible for this are in `browser_protocol.pdl`

File: front_end/generated/protocol.d.ts (hunk 4 above)
Comment (palikar) [reply front_end/generated/protocol.d.ts:13007]: this is a generated file. All `COHERENT BEGIN` / `COHERENT END` comments for changes responsible for this are in `browser_protocol.pdl`

File: front_end/panels/timeline/TimelinePanel.ts (hunk 5 above)
Comment (palikar) [reply front_end/panels/timeline/TimelinePanel.ts]: sure, I reworked the tracing agent acquisition logic based on the suggestion and it seems fine. I've tested the new logic and everything works correctly (refreshing during tracing leads to no crashes) .

**Review by omalinov — APPROVED** [review]

**Pull request merged**